### PR TITLE
[Documentation] Documentation for loss functions.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#1004

*Description of changes:*
Add documentation for loss functions.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by classicsong** [comment]
https://graphstorm-xiang.readthedocs.io/en/loss-func-doc/api/references/index.html
Here is the doc pre-view

File: python/graphstorm/model/loss_func.py [hunk 1]
```diff
@@ -363,19 +477,38 @@ def forward(self, pos_score, neg_score):
         return loss
 
 class WeightedLinkPredictAdvBCELossFunc(LinkPredictAdvBCELossFunc):
-    """ BCE loss function for link prediction with adversarial
-        loss for negative samples and weight on positive samples.
+    r""" Binary cross entropy loss function for link prediction tasks
+    with adversarial loss for negative samples and weight on positive samples.
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:481]:
```suggestion
    with adversarial loss for negative samples and weights on positive samples.
```

File: python/graphstorm/model/loss_func.py [hunk 2]
```diff
@@ -363,19 +477,38 @@ def forward(self, pos_score, neg_score):
...
+
+    The loss function of a positive edge ``e`` is as:
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:483]:
```suggestion
    The loss function of a positive edge is defined as:
```

File: python/graphstorm/model/loss_func.py [hunk 3]
```diff
@@ -363,19 +477,38 @@ def forward(self, pos_score, neg_score):
...
+
+    .. math::
+
+        loss_{pos} = - w * \log score
+
+    where ``score`` is the score value of the positive edges computed by the score function, ``w`` is the weight of each positive edge. The loss of the negative edges is the same as **LinkPredictAdvBCELossFunc**.
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:489]:
```suggestion
    where ``score`` is the score value of the positive edges computed by the score function, and ``w`` is the weight of each positive edge. The loss of the negative edges is the same as **LinkPredictAdvBCELossFunc**.
```

File: python/graphstorm/model/loss_func.py [hunk 4]
```diff
@@ -363,19 +477,38 @@ def forward(self, pos_score, neg_score):
...
+
+    The final loss is as:
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:491]:
```suggestion
    The final loss is defined as:
```

File: python/graphstorm/model/loss_func.py [hunk 5]
```diff
@@ -412,29 +545,32 @@ def forward(self, pos_score, neg_score):
 class LinkPredictContrastiveLossFunc(GSLayer):
     r""" Contrastive Loss function for link prediction.
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:546]:
```suggestion
    r""" Contrastive Loss function for link prediction tasks.
```

File: python/graphstorm/model/loss_func.py [hunk 6]
```diff
@@ -412,29 +545,32 @@ def forward(self, pos_score, neg_score):
...
+    The positive and negative scores are computed through a
+    score function as:
+
+        .. math::
 
             score = f(<src, rel, dst>)
 
-        And we treat a score as the distance between `src` and
-        `dst` nodes under relation `rel`.
+    And we treat a score as the distance between `src` and
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:555]:
```suggestion
    And we treat a score as the distance between ``src`` and
```

File: python/graphstorm/model/loss_func.py [hunk 7]
```diff
@@ -412,29 +545,32 @@ def forward(self, pos_score, neg_score):
...
+    `dst` nodes under relation `rel`.
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:556]:
```suggestion
    ``dst`` nodes under relation ``rel``.
```

File: python/graphstorm/model/loss_func.py [hunk 8]
```diff
@@ -412,29 +545,32 @@ def forward(self, pos_score, neg_score):
...
 
-        In contrastive loss, we assume one positive pair <src, dst>
-        has K corresponding negative pairs <src, neg_dst1>,
-        <src, neg_dst2> .... <src, neg_dstk> When we compute the
-        loss of <src, dst>, we follow the following equation:
+    In contrastive loss, we assume one positive pair <src, dst>
+    has K corresponding negative pairs <src, neg_dst1>,
+    <src, neg_dst2> .... <src, neg_dstk> When we compute the
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:560]:
```suggestion
    <src, neg_dst2> .... <src, neg_dstk>. When we compute the
```

File: python/graphstorm/model/loss_func.py [hunk 9]
```diff
@@ -412,29 +545,32 @@ def forward(self, pos_score, neg_score):
...
+
+        .. math::
 
-            .. math::
             loss = -log(exp(pos\_score)/\sum_{i=0}^N exp(score_i))
 
-        where score includes both positive score of <src, dst> and
-        negative scores of <src, neg_dst0>, ... <src, neg_dstk>
+    where score includes both positive score of <src, dst> and
+    negative scores of <src, neg_dst0>, ... <src, neg_dstk>
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:568]:
```suggestion
    negative scores of <src, neg_dst0>, ... <src, neg_dstk>.
```

File: python/graphstorm/model/loss_func.py [hunk 10]
```diff
@@ -412,29 +545,32 @@ def forward(self, pos_score, neg_score):
...
 
-        Parameters
-        ----------
-        temp: float
-            Temperature value
+    Parameters
+    ----------
+    temp: float
+        Temperature value
```
Comment (zhjwy9343) [inline python/graphstorm/model/loss_func.py:573]:
```suggestion
        Temperature value.
```

File: python/graphstorm/model/loss_func.py (hunk 9 above)
Comment (zhjwy9343) [reply python/graphstorm/model/loss_func.py:568]: Is the ``N`` in the equation same as the ``k`` of neg edge pairs? If so, better to change ``N`` to ``k``.

File: python/graphstorm/model/loss_func.py [hunk 11]
```diff
@@ -222,18 +286,44 @@ def out_dims(self):
...
+    .. math::
+
+        w\_e = \left \{
+        \begin{array}{lc}
+            1,  & \text{ if } e \in G, \\
+            0,  & \text{ if } e \notin G
+        \end{array}
+        \right.
+
+    where ``G`` is the training graph.
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:310]: ``e`` is mentioned in L299

File: python/graphstorm/model/loss_func.py [hunk 12]
```diff
@@ -279,17 +369,36 @@ def out_dims(self):
...
-            Temperature value for adversarial loss.
+        loss_{pos} = - \log score
+
+    where ``score`` is the score value of the positive edges computed by the score function.
+
+    The loss of negative edges is as:
+
+    .. math::
+
+        loss_{neg} = \log (1 - score) \\
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:387]: I just split the equation into two lines.

[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:387]: L387 is to compute the original loss_neg
and L388 is to compute the adversarial loss_neg

File: docs/source/api/references/index.rst [hunk 13]
```diff
@@ -5,7 +5,7 @@ API Reference
 
 The released APIs include the necessary components that can build GraphStorm models, training and inference pipelines. For examples about how to use these APIs, please refer to the :ref:`API Programming Examples<programming-examples>`.
 
-.. note:: This API reference does not include the components for multi-task learning, which is still an experimental feature of Graphstorm v0.3, and may need extensive modification. 
+.. note:: This API reference does not include the components for multi-task learning, which is still an experimental feature of Graphstorm v0.3, and may need extensive modification.
```
Comment (thvasilo) [inline docs/source/api/references/index.rst:8]: Maybe we want to remove the mention of the version here until MT is out of experimental?

File: python/graphstorm/model/loss_func.py [hunk 14]
```diff
@@ -142,17 +175,32 @@ def out_dims(self):
         return None
 
 class RegressionLossFunc(GSLayer):
-    """ Loss function for regression
+    """ Loss function for regression tasks.
+
+    The the ``torch.nn.MSELoss`` is used.
```
Comment (thvasilo) [inline python/graphstorm/model/loss_func.py:180]:
```suggestion
    The ``torch.nn.MSELoss`` is used.
```

File: python/graphstorm/model/loss_func.py [hunk 15]
```diff
@@ -175,18 +223,34 @@ def out_dims(self):
...
+
+    The ``torch.nn.functional.binary_cross_entropy_with_logits`` is used
+    to compute the loss. The loss function is defined as:
+
+    .. math::
+
+        loss = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
+
+    where ``y`` is 1 for a positive edge and 0 for a negative edge.
+    ``score`` is the score value of ``e`` computed by the score function.
```
Comment (thvasilo) [inline python/graphstorm/model/loss_func.py:237]: What does `e` refer to here? The edge?

Comment (zhjwy9343) [reply python/graphstorm/model/loss_func.py:237]: I have the same confusion.

File: python/graphstorm/model/loss_func.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/loss_func.py:313]: In L299 (current L302), we didn't mention that ``e`` represents an edge.

File: python/graphstorm/model/loss_func.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/loss_func.py:391]: I see. But this expression is not like a Math function, but a programming logic. If possible, we can put them in one formula.

File: python/graphstorm/model/loss_func.py (hunk 15 above)
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:237]: Updated

File: docs/source/api/references/index.rst (hunk 13 above)
Comment (classicsong) [reply docs/source/api/references/index.rst:8]: We can update it later. Maybe we can remove it from experimental in 0.4

**Review by zhjwy9343 — APPROVED** [review]
LGTM
